src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1640,55 +1751,150 @@ public void ReturnConnection(HttpConnection connection)
         {
+            if (NetEventSource.Log.IsEnabled()) connection.Trace($"{nameof(isNewConnection)}={isNewConnection}");
+
+            if (!isNewConnection && CheckExpirationOnReturn(connection))
+            {
+                lock (SyncObj)
+                {
+                    Debug.Assert(_availableHttp2Connections is null || !_availableHttp2Connections.Contains(connection));
+                    Debug.Assert(_associatedHttp2ConnectionCount > (_availableHttp2Connections?.Count ?? 0));
+                    _associatedHttp2ConnectionCount--;
+                }
+
+                if (NetEventSource.Log.IsEnabled()) connection.Trace("Disposing HTTP/2 connection return to pool. Connection lifetime expired.");
+                connection.Dispose();
+                return;
+            }
+
+            bool usable = true;
+            bool poolDisposed = false;
             lock (SyncObj)
             {
-                Http2Connection[]? localHttp2Connections = _http2Connections;
+                Debug.Assert(_availableHttp2Connections is null || !_availableHttp2Connections.Contains(connection));
+                Debug.Assert(_associatedHttp2ConnectionCount > (_availableHttp2Connections?.Count ?? 0));
 
-                if (localHttp2Connections == null)
+                if (isNewConnection)
                 {
-                    return;
+                    Debug.Assert(_pendingHttp2Connection);
+                    _pendingHttp2Connection = false;
                 }
 
-                if (localHttp2Connections.Length == 1)
+                while (!_http2RequestQueue.IsEmpty)
                 {
-                    // Fast shortcut for the most common case.
-                    if (localHttp2Connections[0] == connection)
+                    Debug.Assert((_availableHttp2Connections?.Count ?? 0) == 0, $"With {_availableHttp11Connections.Count} available HTTP2 connections, we shouldn't have a waiter.");
+
+                    if (!connection.TryReserveStream())
                     {
-                        _http2Connections = null;
+                        usable = false;
+                        if (isNewConnection)
+                        {
+                            // The new connection could not handle even one request, either because it shut down before we could use it for any requests,
+                            // or because it immediately set the max concurrent streams limit to 0.
+                            // We don't want to get stuck in a loop where we keep trying to create new connections for the same request.
+                            // Fail the next request, if any.
+                            HttpRequestException hre = new HttpRequestException(SR.net_http_http2_connection_not_established);
+                            ExceptionDispatchInfo.SetCurrentStackTrace(hre);
+                            _http2RequestQueue.TryFailNextRequest(hre);
```

Comment:
I'm not sure if fully understand how this approach handles multiple HTTP/2 connections. `_http2RequestQueue` is shared by all connection, so it looks like the next request dequeuing can happen on a different connection that the one passed to here, therefore that other connection will be filed by this exception.
Am I missing something? Could you please clarify the logic here?

Reply:
Not sure I understand the question.

TryFailNextRequest will dequeue the request and fail it. Is that what you're asking about? I can change the name to be clearer. It's meant to parallel TryDequeueNextRequest below, except that the request will be failed with the specific exception.